### PR TITLE
Express notes api

### PR DESCRIPTION
**Pull request by daniel-sungwon-lee**
express-notes-api

### CODE REVIEW

File: express-notes-api/index.js [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+  } else {
+    const err = {
+      "error": "id must be a positive integer"
+    }
+    res.status(400).json(err)
+  }
+})
+
+
+app.use(express.json())
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: This should be above all of the endpoints

File: express-notes-api/index.js [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  for (const prop in notes){
+    notesArr.push(notes[prop])
+  }
+
+  res.status(200).json(notesArr)
+})
+
+app.get("/api/notes/:id",(req,res)=>{
+  if (isNaN(req.params.id) === false){
+    let id = parseFloat(req.params.id)
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: Number must be a positive integer, parseInt is more appropriate here.

File: express-notes-api/index.js [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+
+    const idArr=[]
+    for (const prop in notes) {
+      idArr.push(parseInt(prop))
+    }
+
+    if (Number.isInteger(id)){
+      if(Math.sign(id)===1){
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: These two lines would be unnecessary if you modified line 21 to also check is `req.params.id` is greater than zero. Less checks = happy code

File: express-notes-api/index.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+      const err = {
+        "error": "id must be a positive integer"
+      }
+      res.status(400).json(err)
+    }
+  } else {
+    const err = {
+      "error": "id must be a positive integer"
+    }
+    res.status(400).json(err)
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: I would refactor these to reduce the amount of times you are checking the same thing in the data

File: express-notes-api/index.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+  }
+})
+
+
+app.use(express.json())
+
+app.post("/api/notes",(req,res)=>{
+  let newNote=req.body
+
+  if (Object.keys(newNote).length===0){
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: you only need to check if the object has a content property

File: express-notes-api/index.js [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    const err = {
+      "error": "content is a required field"
+    }
+
+    res.status(400).json(err)
+
+  } else {
+    newNote.id=parseInt(nextId)
+
+    notes[nextId]=newNote
```
Comment (BrettAlbright) [inline express-notes-api/index.js:56]: this needs to be specifically the content property of the req.body object

File: express-notes-api/index.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+app.delete("/api/notes/:id",(req,res)=>{
+  if (isNaN(req.params.id)===false){
+    let id = parseFloat(req.params.id)
+
+    const idArr = []
+    for (const prop in notes) {
+      idArr.push(parseInt(prop))
+    }
+
+    if (Number.isInteger(id)) {
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: like above, these two checks can be removed if you slightly modify line 93

File: express-notes-api/index.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        "error": "id must be a positive integer"
+      }
+      res.status(400).json(err)
+    }
+  } else {
+    const err = {
+      "error": "id must be a positive integer"
+    }
+    res.status(400).json(err)
+  }
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: please refactor and simplify. there should only be one cast of an error for id must be a positive integer

File: express-notes-api/index.js [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  } else {
+    if (isNaN(req.params.id) === false){
+      let id = parseFloat(req.params.id)
+
+      const idArr = []
+      for (const prop in notes) {
+        idArr.push(parseInt(prop))
+      }
+
+      if (Number.isInteger(id)) {
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: same as noted before

File: express-notes-api/index.js [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+    }
+    res.status(400).json(err)
+  }
+})
+
+
+app.put("/api/notes/:id",(req,res)=>{
+  let updatedNote = req.body
+
+  if (Object.keys(updatedNote).length === 0) {
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: you need to check specifically for the content property

File: express-notes-api/index.js [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+
+      if (Number.isInteger(id)) {
+        if (Math.sign(id) === 1) {
+          if (idArr.includes(id)) {
+            updatedNote.id = id
+            notes[id] = updatedNote
+
+            fs.writeFile("data.json", JSON.stringify(data, null, 2), (err) => {
+              if (err) {
+                console.error(err)
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: this should not be a console log, it should send back a 500 error

File: express-notes-api/index.js [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    if (Number.isInteger(id)) {
+      if (Math.sign(id) === 1) {
+        if (idArr.includes(id)) {
+          delete notes[id]
+
+          fs.writeFile("data.json", JSON.stringify(data, null, 2), (err) => {
+            if (err) {
+              console.error(err)
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: This should not be a console log, it should send back a 500 status to the client

**Review by Zhiwutian — CHANGES_REQUESTED** [review]
some updates for you good sir!

File: express-notes-api/index.js [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+    const err = {
+      "error": "id must be a positive integer"
+    }
+    res.status(400).json(err)
+  }
+})
+
+
+app.post("/api/notes",(req,res)=>{
+  let newNote=req.body
```
Comment (BrettAlbright) [inline express-notes-api/index.js]: here you should be using req.body.content. you can't guarantee that the body will just be a single property

**Review by Zhiwutian — CHANGES_REQUESTED** [review]
tiny update!

**Comment by daniel-sungwon-lee** [comment]
image
image

**Review by Zhiwutian — APPROVED** [review]
👍

**Pull request merged**